src/test/recovery/t/139_archive_mode_always.pl

```diff
@@ -0,0 +1,79 @@
+# test WAL persistance for archive_mode 'always'
+use strict;
+use warnings;
+use PostgreSQL::Test::Cluster;
+use Test::More tests => 2;
+use File::Copy;
+
+# The idea of the test is to check that not archived WAL files are not removed
+# by creation of checkpoint (on master) or restartpoint (on standby).
+
+# Initialize master node, doing archives
+my $node_master = PostgreSQL::Test::Cluster->new('master');
+$node_master->init(
+	has_archiving    => 1,
+	allows_streaming => 1);
+my $backup_name = 'my_backup';
+
+# Append necessary config. Some of params override already initialized ones.
+# archive_mode 'always' is a subject of the test.
+# archive_command 'false' is used to prevent archiving. The idea is to check
+#   WAL file is not removed by checkpointer. Successfull archiving (with file
+#   removing) should not affect the test.
+# wal_keep_size '0' is used to prevent keeping old WAL files. As test
+#   checks that WAL file still exist, it should not be kept by any other
+#   mechanics.
+# create_restartpoint_on_ckpt_record_replay 'on' makes restartpoint (and
+#   waiting for completion) while replaying checkpoint record on standby.
+$node_master->append_conf(
+		'postgresql.conf', qq{
+archive_mode = 'always'
+archive_command = '/usr/bin/false'
+wal_keep_size = 0
+create_restartpoint_on_ckpt_record_replay = 'on'
+});
+
+# Start it
+$node_master->start;
+
+# Take backup for standby
+$node_master->backup($backup_name);
+
+# Initialize standby node from backup, fetching WAL from archives
+my $node_standby = PostgreSQL::Test::Cluster->new('standby');
+$node_standby->init_from_backup($node_master, $backup_name,
+	has_streaming => 1);
+$node_standby->start;
+
+# Create some content on master
+$node_master->safe_psql('postgres',
+	"CREATE TABLE t AS SELECT generate_series(1,1000) i;");
+
+# Find the next WAL segment to be archived
+my $walfile_to_be_kept = $node_master->safe_psql('postgres',
+	"SELECT pg_walfile_name(pg_current_wal_lsn());");
+
+# Switch WAL and do checkpoint.
+# Checkpoint is requested with CHECKPOINT_WAIT, so it's enough to wait for
+# command end.
+$node_master->safe_psql('postgres', "
+SELECT pg_switch_wal();
+CHECKPOINT;
+");
+
+my $current_lsn =
+  $node_master->safe_psql('postgres', "SELECT pg_current_wal_lsn();");
```

Comment:
Is it more correct to use pg_current_wal_insert_lsn?

Reply:
From postgres documentation:

"`pg_current_wal_lsn` displays the current write-ahead log write location.
`pg_current_wal_insert_lsn` displays the current write-ahead log insertion location.
The insertion location is the “logical” end of the write-ahead log at any instant,
while the write location is the end of what has actually been written out from the server's internal buffers.
The write location is the end of what can be examined from outside the server, and is usually what you want
if you are interested in archiving partially-complete write-ahead log files."

So, `pg_current_wal_insert_lsn()` is the internal buffer head location.
`pg_current_wal_lsn()` is what had been written out to kernel.

Currently `pg_current_wal_lsn()` is used in 2 places:
1. to find the next WAL segment to be archived after pushing some content to the master. Here `pg_current_wal_insert_lsn()` may be more suitable, as by the moment of sampling the `$walfile_to_be_kept` we can't guarantee that all internal buffers were pushed to the kernel.
2. to wait for the moment of replay had been done on standby. Here `pg_current_wal_lsn()` looks Ok.